File: src/torchcodec/_core/Logging.cpp
// Copyright (c) Meta Platforms, Inc. and affiliates.
// All rights reserved.
//
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree.

#include "Logging.h"

#include <atomic>
#include <cstdarg>
#include <cstdio>
#include <filesystem>

namespace facebook::torchcodec {

static std::atomic<int> g_log_level{static_cast<int>(LogLevel::OFF)};

void set_cpp_log_level(LogLevel level) {
  g_log_level.store(static_cast<int>(level), std::memory_order_relaxed);
}

LogLevel get_log_level() {
  return static_cast<LogLevel>(g_log_level.load(std::memory_order_relaxed));
}

namespace internal {

void log(const char* file, int line, const char* fmt, ...) {
  auto basename = std::filesystem::path(file).filename().string();

  std::fprintf(stderr, "[torchcodec %s:%d] ", basename.c_str(), line);

  va_list args;
  va_start(args, fmt);
  std::vfprintf(stderr, fmt, args);
  va_end(args);

  std::fprintf(stderr, "\n");
}

} // namespace internal
} // namespace facebook::torchcodec

File: src/torchcodec/_core/_logging.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
#
# This source code is licensed under the BSD-style license found in the
# LICENSE file in the root directory of this source tree.

import logging

from torchcodec._core.ops import _get_log_level, _set_log_level

_LG = logging.getLogger("torchcodec")
_LG.setLevel(logging.DEBUG)

_handler = logging.StreamHandler()
_handler.setFormatter(
    logging.Formatter("[torchcodec %(filename)s:%(lineno)d] %(message)s")
)

# Maps user-facing level names to C++ log level ints.
# Currently only OFF and ALL are supported; more levels will be added later.
_LEVEL_NAME_TO_INT = {
    "OFF": 0,
    "ALL": 1,
}

_LEVEL_INT_TO_NAME = {v: k for k, v in _LEVEL_NAME_TO_INT.items()}


def set_log_level(level: str) -> None:
    """Set the torchcodec log level (both Python and C++ sides).

    Supported levels: "OFF", "ALL".
    """
    level = level.upper()
    if level not in _LEVEL_NAME_TO_INT:
        raise ValueError(
            f"Invalid log level: {level!r}. Supported levels: {list(_LEVEL_NAME_TO_INT.keys())}"
        )
    level_int = _LEVEL_NAME_TO_INT[level]
    _set_log_level(level_int)
    if level_int > 0:
        if _handler not in _LG.handlers:
            _LG.addHandler(_handler)
    else:
        _LG.removeHandler(_handler)


def get_log_level() -> str:
    """Return the current torchcodec log level as a string."""
    level_int = _get_log_level()
    return _LEVEL_INT_TO_NAME.get(level_int, f"UNKNOWN({level_int})")
